crates/ty_ide/src/rename.rs

```diff
@@ -0,0 +1,773 @@
+use crate::goto::find_goto_target;
+use crate::references::{ReferencesMode, references};
+use crate::{Db, ReferenceTarget};
+use ruff_db::files::File;
+use ruff_text_size::{Ranged, TextSize};
+use ty_python_semantic::ImportAliasResolution;
+
+/// Returns the range of the symbol if it can be renamed, None if not.
+pub fn can_rename(db: &dyn Db, file: File, offset: TextSize) -> Option<ruff_text_size::TextRange> {
+    let parsed = ruff_db::parsed::parsed_module(db, file);
+    let module = parsed.load(db);
+
+    // Get the definitions for the symbol at the offset
+    let goto_target = find_goto_target(&module, offset)?;
+
+    // Don't allow renaming of import module components
+    if matches!(
+        goto_target,
+        crate::goto::GotoTarget::ImportModuleComponent { .. }
+    ) {
+        return None;
+    }
+
+    let current_file_in_project = is_file_in_project(db, file);
+
+    if let Some(definition_targets) =
+        goto_target.get_definition_targets(file, db, None, ImportAliasResolution::PreserveAliases)
+    {
+        for target in &definition_targets {
+            let target_file = target.file();
+
+            // If definition is outside the project, refuse rename
+            if !is_file_in_project(db, target_file) {
+                return None;
+            }
+
+            // If current file is not in project and any definition is outside current file, refuse rename
+            if !current_file_in_project && target_file != file {
+                return None;
+            }
+        }
+    } else {
+        // No definition targets found. This happens for keywords, so refuse rename
+        return None;
+    }
+
+    Some(goto_target.range())
+}
+
+/// Perform a rename operation on the symbol at the given position.
+/// Returns all locations that need to be updated with the new name.
+pub fn rename(
+    db: &dyn Db,
+    file: File,
+    offset: TextSize,
+    new_name: &str,
+) -> Option<Vec<ReferenceTarget>> {
+    let parsed = ruff_db::parsed::parsed_module(db, file);
+    let module = parsed.load(db);
+
+    // Get the definitions for the symbol at the offset
+    let goto_target = find_goto_target(&module, offset)?;
+
+    // Clients shouldn't call us with an empty new name, but just in case...
+    if new_name.is_empty() {
```

Comment:
Do we need to perform any validation on whether the new name is a valid python name?